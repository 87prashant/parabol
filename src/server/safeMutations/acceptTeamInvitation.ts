import adjustUserCount from 'server/billing/helpers/adjustUserCount'
import getRethink from 'server/database/rethinkDriver'
import addTeamMemberToNewMeeting from 'server/graphql/mutations/helpers/addTeamMemberToNewMeeting'
import addUserToTMSUserOrg from 'server/safeMutations/addUserToTMSUserOrg'
import insertNewTeamMember from 'server/safeMutations/insertNewTeamMember'
import {ADD_USER} from 'server/utils/serverConstants'
import {TEAM_INVITATION} from 'universal/utils/constants'
import {auth0ManagementClient} from 'server/utils/auth0Helpers'

const acceptTeamInvitation = async (
  teamId: string,
  userId: string,
  invitationId: string,
  dataLoader: any
) => {
  const r = getRethink()
  const now = new Date()
  const {
    team: {orgId},
    user
  } = await r({
    team: r.table('Team').get(teamId),
    user: r.table('User').get(userId)
  })
  const userOrgs = user.userOrgs || []
  const userInOrg = Boolean(userOrgs.find((org) => org.id === orgId))
  const {removedNotificationIds, teamMember} = await r({
    // add the team to the user doc
    userUpdate: addUserToTMSUserOrg(userId, teamId, orgId),
    teamMember: insertNewTeamMember(userId, teamId),
    // only redeem 1 invitation. any others will expire
    redeemedInvitation: r
      .table('TeamInvitation')
      .get(invitationId)
      .update({
        acceptedAt: now,
        acceptedBy: userId
      }),
    removedNotificationIds: r
      .table('Notification')
      .getAll(userId, {index: 'userIds'})
      .filter({
        type: TEAM_INVITATION,
        teamId
      })
      .update({isArchived: true}, {returnChanges: true})('changes')('new_val')('id')
  })
  if (!userInOrg) {
    await adjustUserCount(userId, orgId, ADD_USER)
  }

  // if a meeting is going on right now, add them
  await addTeamMemberToNewMeeting(teamMember, teamId, dataLoader)

  // update auth0
  const tms = user.tms ? user.tms.concat(teamId) : [teamId]
  auth0ManagementClient.users.updateAppMetadata({id: userId}, {tms})
  return removedNotificationIds
}

export default acceptTeamInvitation
