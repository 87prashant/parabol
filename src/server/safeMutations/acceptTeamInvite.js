import adjustUserCount from 'server/billing/helpers/adjustUserCount';
import getRethink from 'server/database/rethinkDriver';
import addUserToTMSUserOrg from 'server/safeMutations/addUserToTMSUserOrg';
import insertNewTeamMember from 'server/safeMutations/insertNewTeamMember';
import getTeamInviteNotifications from 'server/safeQueries/getTeamInviteNotifications';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {getUserId} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import {ADD_USER} from 'server/utils/serverConstants';
import tmsSignToken from 'server/utils/tmsSignToken';
import {
  ADD_TO_TEAM, ADDED, INVITATION_REMOVED, JOIN_TEAM, NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED, NOTIFICATIONS_CLEARED,
  TEAM_ADDED, TEAM_MEMBER
} from 'universal/utils/constants';
import toTeamMemberId from 'universal/utils/relay/toTeamMemberId';

const publishRemovedInvitations = (expiredEmailInvitations, operationId) => {
  expiredEmailInvitations.forEach((invitation) => {
    const {teamId} = invitation;
    const invitationRemoved = {invitation};
    getPubSub().publish(`${INVITATION_REMOVED}.${teamId}`, {invitationRemoved, operationId});
  });
};

const publishClearNotifications = (expireInviteNotificationIds, userId) => {
  if (expireInviteNotificationIds.length > 0) {
    const notificationsCleared = {deletedIds: expireInviteNotificationIds};
    getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${userId}`, {notificationsCleared});
  }
};

const publishJoinTeamNotifications = (teamId, teamName, user) => {
  const notificationsAdded = {
    notifications: [{
      type: JOIN_TEAM,
      teamId,
      teamName,
      preferredName: user.preferredName || user.email
    }]
  };
  getPubSub().publish(`${NOTIFICATIONS_ADDED}.${teamId}`, {notificationsAdded});
};

const acceptTeamInvite = async (teamId, authToken, email, subOptions = {}) => {
  const r = getRethink();
  const now = new Date();
  const {operationId} = subOptions;
  const userId = getUserId(authToken);
  const teamMemberId = toTeamMemberId(teamId, userId);
  const {team: {orgId, name: teamName}, user} = await r({
    team: r.table('Team').get(teamId).pluck('orgId', 'name'),
    user: r.table('User').get(userId)
  });
  const userOrgs = user.userOrgs || [];
  const userTeams = user.tms || [];
  const userInOrg = Boolean(userOrgs.find((org) => org.id === orgId));
  const tms = [...userTeams, teamId];
  const {expiredEmailInvitations, expireInviteNotificationIds, teamMember} = await r({
    // add the team to the user doc
    userUpdate: addUserToTMSUserOrg(userId, teamId, orgId),
    teamMember: insertNewTeamMember(userId, teamId),
    // find all possible emails linked to this person and mark them as accepted
    expiredEmailInvitations: r.table('Invitation')
      .getAll(email, {index: 'email'})
      .update({
        acceptedAt: now,
        // flag the token as expired so they cannot reuse the token
        tokenExpiration: new Date(0),
        updatedAt: now
      }, {returnChanges: true})('changes')('new_val').default([]),
    expireInviteNotificationIds: getTeamInviteNotifications(orgId, teamId, [email])
      .delete({returnChanges: true})('changes')('old_val')('id')
      .default([])
  });

  if (!userInOrg) {
    await adjustUserCount(userId, orgId, ADD_USER);
  }
  auth0ManagementClient.users.updateAppMetadata({id: userId}, {tms});

  publishRemovedInvitations(expiredEmailInvitations, operationId);
  publishClearNotifications(expireInviteNotificationIds, userId);
  publishJoinTeamNotifications(teamId, teamName, user);

  getPubSub().publish(`${TEAM_MEMBER}.${teamId}`, {teamMember: {teamMemberId, type: ADDED}, ...subOptions});

  // Send the new team member a welcome & a new token
  const newAuthToken = tmsSignToken(authToken, tms);
  const addedToTeam = {
    // no ID to rule out permanent notification. they already know, they are the ones who did it!
    startAt: now,
    type: ADD_TO_TEAM,
    teamId
  };
  const teamAdded = {
    notification: addedToTeam,
    teamId,
    teamMemberId
  };
  getPubSub().publish(`${TEAM_ADDED}.${userId}`, {teamAdded, ...subOptions});
  getPubSub().publish(`${NEW_AUTH_TOKEN}.${userId}`, {newAuthToken});

  return {
    authToken: newAuthToken,
    teamId,
    teamMemberId
  };
};

export default acceptTeamInvite;
