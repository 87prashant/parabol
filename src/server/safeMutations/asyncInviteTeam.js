import getRethink from 'server/database/rethinkDriver';
import getInviterInfoAndTeamName from '../graphql/models/Invitation/inviteTeamMembers/getInviterInfoAndTeamName';
import createEmailPromises from '../graphql/models/Invitation/inviteTeamMembers/createEmailPromises';
import resolveSentEmails from '../graphql/models/Invitation/inviteTeamMembers/resolveSentEmails';
import makeInviteToken from '../graphql/models/Invitation/inviteTeamMembers/makeInviteToken';
import makeInvitationsForDB from '../graphql/models/Invitation/inviteTeamMembers/makeInvitationsForDB';

export default async function asyncInviteTeam(inviterUserId, teamId, invitees) {
  if (invitees.length === 0) return invitees;
  const r = getRethink();
  const inviteesWithTokens = invitees.map((invitee) => ({...invitee, inviteToken: makeInviteToken()}));
  const inviterInfoAndTeamName = await getInviterInfoAndTeamName(teamId, inviterUserId);
  const sendEmailPromises = createEmailPromises(inviterInfoAndTeamName, inviteesWithTokens);
  const {inviteesToStore} = await resolveSentEmails(sendEmailPromises, inviteesWithTokens);
  const invitationsForDB = await makeInvitationsForDB(inviteesToStore, teamId, inviterUserId);
  // Bulk insert, wait in case something queries the invitation table
  await r.table('Invitation').insert(invitationsForDB);
  return inviteesWithTokens;
  // TODO generate email to inviter including folks that we couldn't reach
  // if (inviteeErrors.length > 0) {
  // throw errorObj({_error: 'Some invitations were not sent', type: 'inviteSendFail', failedEmails: inviteeErrors});
  // }
};
