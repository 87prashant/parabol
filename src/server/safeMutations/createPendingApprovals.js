import getRethink from 'server/database/rethinkDriver'
import {PENDING} from 'server/utils/serverConstants'
import shortid from 'shortid'
import {BILLING_LEADER, REQUEST_NEW_USER} from 'universal/utils/constants'
import makeNewSoftTeamMembers from 'server/safeMutations/makeNewSoftTeamMembers'

export default async function createPendingApprovals (outOfOrgEmails, inviteSender, dataLoader) {
  if (outOfOrgEmails.length === 0) {
    return {
      requestNotifications: [],
      orgApprovalIds: [],
      billingLeaderUserIds: [],
      newSoftTeamMembers: []
    }
  }
  const {orgId, teamId, userId} = inviteSender || {}
  const r = getRethink()
  const now = new Date()
  // add a notification to the Billing Leaders
  const {userIds, inviter} = await r({
    userIds: r
      .table('User')
      .getAll(orgId, {index: 'userOrgs'})
      .filter((user) =>
        user('userOrgs').contains((userOrg) =>
          userOrg('id')
            .eq(orgId)
            .and(userOrg('role').eq(BILLING_LEADER))
        )
      )('id')
      .coerceTo('array'),
    inviter: r
      .table('User')
      .get(userId)
      .pluck('preferredName', 'id')
  })
  const requestNotifications = outOfOrgEmails.map((inviteeEmail) => ({
    id: shortid.generate(),
    type: REQUEST_NEW_USER,
    startAt: now,
    orgId,
    userIds,
    inviterUserId: inviter.id,
    inviteeEmail,
    teamId
  }))

  const pendingApprovals = outOfOrgEmails.map((inviteeEmail) => ({
    id: shortid.generate(),
    createdAt: now,
    email: inviteeEmail,
    isActive: true,
    orgId,
    status: PENDING,
    teamId,
    updatedAt: now
  }))

  // send a new notification to each Billing Leader concerning each out-of-org invitee
  await r({
    requestNotifications: r.table('Notification').insert(requestNotifications),
    approvals: r.table('OrgApproval').insert(pendingApprovals)
  })

  const newSoftTeamMembers = await makeNewSoftTeamMembers(outOfOrgEmails, teamId, dataLoader)

  return {
    requestNotifications,
    newSoftTeamMembers,
    orgApprovalIds: pendingApprovals.map(({id}) => id),
    billingLeaderUserIds: userIds
  }
}
