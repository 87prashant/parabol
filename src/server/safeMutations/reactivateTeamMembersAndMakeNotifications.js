import getRethink from 'server/database/rethinkDriver';
import shortid from 'shortid';
import {ADD_TO_TEAM} from 'universal/utils/constants';
import makeReactivationNotifications from 'server/safeMutations/helpers/makeReactivationNotifications';

const reactivateTeamMembersAndMakeNotifications = async (invitees, inviter, teamMembers) => {
  if (invitees.length === 0) return [];
  const {orgId, teamId, teamName} = inviter;
  const r = getRethink();
  const now = new Date();
  const userIds = invitees.map(({userId}) => userId);
  const teamMemberIds = userIds.map((userId) => `${userId}::${teamId}`);
  const {reactivatedUsers} = await r({
    reactivatedTeamMembers: r.table('TeamMember')
      .getAll(r.args(teamMemberIds), {index: 'id'})
      .update({isNotRemoved: true}),
    reactivatedUsers: r.table('User')
      .getAll(r.args(userIds))
      .update((user) => {
        return user.merge({
          tms: user('tms').append(teamId)
        });
      }, {returnChanges: true})('changes')
      .map((change) => change('new_val'))
      .default([])
  });
  const notifications = reactivatedUsers.map((user) => ({
    id: shortid.generate(),
    type: ADD_TO_TEAM,
    startAt: now,
    orgId,
    userIds: [user.id],
    //inviterName,
    teamName
  }));
  await r.table('Notification').insert(notifications);
  return makeReactivationNotifications(reactivatedUsers, teamMembers, inviter.userId);
};

export default reactivateTeamMembersAndMakeNotifications;
