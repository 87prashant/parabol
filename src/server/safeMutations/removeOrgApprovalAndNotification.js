import getRethink from 'server/database/rethinkDriver';
import {APPROVED, DENIED, PENDING} from 'server/utils/serverConstants';
import {REQUEST_NEW_USER} from 'universal/utils/constants';

const removeOrgApprovalAndNotification = async (orgId, maybeEmails, type) => {
  const now = new Date();
  const {approvedBy, deniedBy} = type;
  const status = approvedBy ? APPROVED : DENIED;
  const emails = Array.isArray(maybeEmails) ? maybeEmails : [maybeEmails];
  const r = getRethink();
  const {removedOrgApprovals, removedRequestNotifications} = await r({
    removedOrgApprovals: r.table('OrgApproval')
      .getAll(r.args(emails), {index: 'email'})
      .filter({orgId, status: PENDING})
      .update({
        status,
        approvedBy,
        deniedBy,
        updatedAt: now
      }, {returnChanges: true})('changes')('new_val')
      .default([]),
    removedRequestNotifications: r.table('Notification')
      .getAll(orgId, {index: 'orgId'})
      .filter({
        type: REQUEST_NEW_USER
      })
      .filter((notification) => {
        return r.expr(emails).contains(notification('inviteeEmail'));
      })
      // get the inviterName
      .delete({returnChanges: true})('changes')('old_val')
      .default([])
  });

  return {
    removedOrgApprovals,
    removedRequestNotifications
  };
};

export default removeOrgApprovalAndNotification;
