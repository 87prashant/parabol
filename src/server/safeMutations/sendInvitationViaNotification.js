import getRethink from 'server/database/rethinkDriver';
import shortid from 'shortid';
import {TEAM_INVITE} from 'universal/utils/constants';

const sendInvitationViaNotification = async (invitees, inviter) => {
  if (invitees.length === 0) return [];
  const r = getRethink();
  const now = new Date();
  const {orgId, inviterName, teamId, teamName} = inviter;
  const invitations = invitees.map((invitee) => ({
    id: shortid.generate(),
    type: TEAM_INVITE,
    startAt: now,
    orgId,
    userIds: [invitee.userId],
    inviterName,
    teamId,
    teamName
  }));
  await r.table('Notification').insert(invitations);
  const notificationsToAdd = {};
  invitations.forEach((notification) => {
    notificationsToAdd[notification.userIds[0]] = [notification];
  });
  return notificationsToAdd;
};

export default sendInvitationViaNotification;
