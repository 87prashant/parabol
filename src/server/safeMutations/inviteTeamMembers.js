import getRethink from 'server/database/rethinkDriver';
import getResults from 'server/graphql/mutations/helpers/inviteTeamMembers/getResults';
import makeDetailedInvitations from 'server/graphql/mutations/helpers/inviteTeamMembers/makeDetailedInvitations';
import createPendingApprovals from 'server/safeMutations/createPendingApprovals';
import reactivateTeamMembersAndMakeNotifications from 'server/safeMutations/reactivateTeamMembersAndMakeNotifications';
import removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';
import sendTeamInvitations from 'server/safeMutations/sendTeamInvitations';
import getTeamInviteNotifications from 'server/safeQueries/getTeamInviteNotifications';
import {isBillingLeader} from 'server/utils/authorization';
import publishNotifications from 'server/utils/publishNotifications';
import {ASK_APPROVAL, REACTIVATE, SEND_INVITATION} from 'server/utils/serverConstants';
import mergeObjectsWithArrValues from 'universal/utils/mergeObjectsWithArrValues';
import resolvePromiseObj from 'universal/utils/resolvePromiseObj';

const inviteTeamMembers = async (invitees, teamId, userId) => {
  const r = getRethink();
  const now = Date.now();
  const {name: teamName, orgId} = await r.table('Team').get(teamId).pluck('name', 'orgId');

  const emailArr = invitees.map((invitee) => invitee.email);
  const {
    inviterDoc,
    pendingInvitations,
    pendingApprovals,
    teamMembers,
    users
  } = await r.expr({
    pendingInvitations: r.table('Invitation')
      .getAll(r.args(emailArr), {index: 'email'})
      .filter((invitation) => invitation('tokenExpiration').ge(r.epochTime(now)))('email')
      .coerceTo('array'),
    //pendingNotificationInvitations: getTeamInviteNotifications(orgId, teamId, emailArr).coerceTo('array'),
    pendingApprovals: r.table('OrgApproval')
      .getAll(r.args(emailArr), {index: 'email'})
      .filter({teamId})
      .coerceTo('array'),
    teamMembers: r.table('TeamMember')
      .getAll(teamId, {index: 'teamId'})
      .coerceTo('array'),
    users: r.table('User')
      .getAll(r.args(emailArr), {index: 'email'})
      .coerceTo('array'),
    inviterDoc: r.table('User').get(userId)
      .merge((doc) => ({
        inviterAvatar: doc('picture'),
        inviterEmail: doc('email'),
        inviterName: doc('preferredName')
      }))
      .pluck('email', 'picture', 'preferredName', 'userOrgs')
  });
  const userOrgDoc = inviterDoc.userOrgs.find((userOrg) => userOrg.id === orgId);
  const inviter = {
    ...inviterDoc,
    orgId,
    teamId,
    teamName,
    userId,
    isBillingLeader: isBillingLeader(userOrgDoc)
  };

  const detailedInvitations = makeDetailedInvitations(teamMembers, emailArr, users, pendingApprovals, pendingInvitations, inviter);
  const inviteesToReactivate = detailedInvitations.filter(({action}) => action === REACTIVATE);
  const inviteesToInvite = detailedInvitations.filter(({action}) => action === SEND_INVITATION);
  const inviteesToApprove = detailedInvitations.filter(({action}) => action === ASK_APPROVAL);

  const approvalEmails = inviteesToApprove.map(({email}) => email);
  const approvalsToClear = inviteesToInvite.map(({email}) => email);

  const {reactivations, notificationsToClear, teamInvites, approvals} = await resolvePromiseObj({
    reactivations: reactivateTeamMembersAndMakeNotifications(inviteesToReactivate, inviter, teamMembers),
    notificationsToClear: removeOrgApprovalAndNotification(orgId, approvalsToClear),
    teamInvites: sendTeamInvitations(inviteesToInvite, inviter),
    approvals: createPendingApprovals(approvalEmails, inviter),
  });

  const notificationsToAdd = mergeObjectsWithArrValues(reactivations, teamInvites, approvals);
  publishNotifications({notificationsToAdd, notificationsToClear});
  const results = getResults(detailedInvitations);
  return {results};
};

export default inviteTeamMembers;
