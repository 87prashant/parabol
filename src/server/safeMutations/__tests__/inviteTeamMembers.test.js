import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import MockDB from 'server/__tests__/setup/MockDB';
import {__now} from 'server/__tests__/setup/mockTimes';
import newInvitee from 'server/__tests__/utils/newInvitee';
import * as publishNotifications from 'server/graphql/mutations/helpers/inviteTeamMembers/publishNotifications';
import * as asyncInviteTeam from 'server/safeMutations/asyncInviteTeam';
import * as createPendingApprovals from 'server/safeMutations/createPendingApprovals';
import inviteTeamMembers from 'server/safeMutations/inviteTeamMembers';
import * as reactivateTeamMembersAndMakeNotifications from 'server/safeMutations/reactivateTeamMembersAndMakeNotifications';
import * as removeOrgApprovalAndNotification from 'server/safeMutations/removeOrgApprovalAndNotification';
import * as sendInvitationViaNotification from 'server/safeMutations/sendInvitationViaNotification';

MockDate.set(__now);
console.error = jest.fn();

describe('inviteTeamMembers', () => {
  test('reactivates a team member that has been invited back to the team', async () => {
    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);

    const mockDB = new MockDB();
    await mockDB.init()
      .teamMember(4, {isNotRemoved: false});
    const firstUser = mockDB.db.user[0];
    const invitees = [{email: mockDB.db.user[4].email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const res = await inviteTeamMembers(invitees, teamId, firstUser.id);

    // VERIFY
    expect(reactivateTeamMembersAndMakeNotifications.default).toHaveBeenCalledTimes(1);
    expect(res).toMatchSnapshot();
  });

  test('invites a new person via email', async () => {

    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    await mockDB.init()
    const firstUser = mockDB.db.user[0];
    const invitee = newInvitee();
    const invitees = [{email: invitee.email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const res = await inviteTeamMembers(invitees, teamId, firstUser.id);

    // VERIFY
    expect(asyncInviteTeam.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });

  test('invites a new person via notification', async () => {
    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB.init()
      .newOrg({name: 'Sad Sacks, Inc.'})
      .newTeam({name: 'The Lonely Ones'})
      .newUser({name: 'Leader of One', email: invitee.email})
      .newTeamMember({isLead: true});
    const firstUser = mockDB.db.user[0];
    const invitees = [{email: mockDB.context.user.email}];
    const teamId = mockDB.context.team.id;
    // TEST
    const res = await inviteTeamMembers(invitees, teamId, firstUser.id);

    // VERIFY
    expect(sendInvitationViaNotification.default).toHaveBeenCalledTimes(1);
    expect(removeOrgApprovalAndNotification.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });

  test('asks the billing leader for approval', async () => {
    // SETUP
    asyncInviteTeam.default = jest.fn(() => []);
    sendInvitationViaNotification.default = jest.fn(() => []);
    reactivateTeamMembersAndMakeNotifications.default = jest.fn(() => []);
    removeOrgApprovalAndNotification.default = jest.fn(() => []);
    createPendingApprovals.default = jest.fn(() => []);
    publishNotifications.default = jest.fn(() => []);
    const dynamicSerializer = new DynamicSerializer();
    const mockDB = new MockDB();
    const invitee = newInvitee();
    await mockDB.init();
    const secondUser = mockDB.db.user[1];
    const invitees = [{email: invitee.email}];
    const teamId = mockDB.context.team.id;

    // TEST
    const res = await inviteTeamMembers(invitees, teamId, secondUser.id);

    // VERIFY
    expect(createPendingApprovals.default).toHaveBeenCalledTimes(1);
    const staticRes = dynamicSerializer.toStatic(res, ['results.email']);
    expect(staticRes).toMatchSnapshot();
  });
});
