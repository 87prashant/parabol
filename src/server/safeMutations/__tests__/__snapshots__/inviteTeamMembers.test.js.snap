// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inviteTeamMembers asks the billing leader for approval 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": undefined,
      "result": "PENDING_APPROVAL",
    },
  ],
}
`;

exports[`inviteTeamMembers invites a new person via email 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": undefined,
      "result": "SUCCESS",
    },
  ],
}
`;

exports[`inviteTeamMembers invites a new person via notification 1`] = `
Object {
  "results": Array [
    Object {
      "email": 0,
      "preferredName": "Leader of One",
      "result": "ALREADY_ON_TEAM",
    },
  ],
}
`;

exports[`inviteTeamMembers reactivates a team member that has been invited back to the team 1`] = `
Object {
  "results": Array [
    Object {
      "email": "bruce.banner@avengers.org",
      "preferredName": "Hulk",
      "result": "SUCCESS",
    },
  ],
}
`;
